### PR TITLE
#321 Handle transfer calls (Switch --> PISP)

### PR DESCRIPTION
**Pull request by sridharvoruganti**
**Handle transfer calls (Switch --> PISP)**

- Updated required swagger definitions
- Handle POST /authorizations/
- Handle PUT /thirdPartyRequest/transaction/{ID}
- Handle PUT /thirdPartyRequest/transaction/{ID}/error
- Unit tests
- Will raise one more PR for mojaloop-simulator changes

### CODE REVIEW

File: src/InboundServer/handlers.js [hunk 1]
```diff
@@ -14,6 +14,8 @@
 const util = require('util');
```
Comment (eoln) [inline src/InboundServer/handlers.js:14]: Sridhar - you should add yourself into the manifest

File: src/InboundServer/handlers.js [hunk 2]
```diff
@@ -600,6 +646,57 @@ const putTransfersByIdError = async (ctx) => {
     ctx.response.body = '';
 };
 
+/**
+ * Handles PUT /thirdPartyRequests/transactions/{ID} request. 
+ * This is response to a POST /thirdPartyRequests/transactions request
+ */
+const putThirdPartyReqTransactionsById = async (ctx) => {
```
Comment (eoln) [inline src/InboundServer/handlers.js:653]: can you refactor the code, so there will be no need to copy-paste the test related code?

File: src/InboundServer/handlers.js [hunk 3]
```diff
@@ -600,6 +646,57 @@ const putTransfersByIdError = async (ctx) => {
...
+        const req = {
+            headers: ctx.request.headers,
+            data: ctx.request.body
+        };
+        const res = await ctx.state.cache.set(`callback_${ctx.state.path.params.ID}`, req);
+        ctx.state.logger.log(`Caching callback: ${util.inspect(res)}`);
+    }
+
+    // publish an event onto the cache for subscribers to action
+    const thirdPartyTrxnChannel = ThirdpartyTrxnModelOut.notificationChannel(ctx.state.path.params.ID);
```
Comment (eoln) [inline src/InboundServer/handlers.js:665]: can you refactor the code of getting the channel name and publishing into the more reusable pattern?

File: src/lib/model/InboundThirdpartyTransactionModel.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ *  Models the operations required for inbound third party transaction requests
+ */
+class InboundThirdpartyTransactionModel {
+    constructor(config) {
+        this._cache = config.cache;
+        this._logger = config.logger;
+        this._dfspId = config.dfspId;
+        this._expirySeconds = config.expirySeconds;
```
Comment (eoln) [inline src/lib/model/InboundThirdpartyTransactionModel.js:34]: where is this property used?

File: src/lib/model/InboundThirdpartyTransactionModel.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            wso2Auth: config.wso2Auth
+        });
+
+        this._backendRequests = new BackendRequests({
+            logger: this._logger,
+            backendEndpoint: config.backendEndpoint,
+            dfspId: config.dfspId
+        });
+
+        this._checkIlp = config.checkIlp;
```
Comment (eoln) [inline src/lib/model/InboundThirdpartyTransactionModel.js:55]: where is this property used?

File: src/lib/model/InboundThirdpartyTransactionModel.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        this._ilp = new Ilp({
```
Comment (eoln) [inline src/lib/model/InboundThirdpartyTransactionModel.js:57]: where is this property used?

File: src/lib/model/InboundThirdpartyTransactionModel.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        let mojaloopErrorCode = Errors.MojaloopApiErrorCodes.INTERNAL_SERVER_ERROR;
+        if (err instanceof HTTPResponseError) {
+            const e = err.getData();
+            if (e.res && e.res.body) {
+                try {
+                    const bodyObj = JSON.parse(e.res.body);
+                    mojaloopErrorCode = Errors.MojaloopApiErrorCodeFromCode(`${bodyObj.statusCode}`);
+                }
+                catch (ex) {
+                    // do nothing
```
Comment (eoln) [inline src/lib/model/InboundThirdpartyTransactionModel.js:100]: here can be exception form `JSON.parse` or from `Errors.MojaloopApiErrorCodeFromCode` are you sure `doing nothing` is a good practice?

File: src/lib/model/OutboundThirdpartyTransactionModel.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ *                                                                        *
+ *  This file is made available under the terms of the license agreement  *
+ *  specified in the corresponding source code repository.                *
+ *                                                                        *
+ *  ORIGINAL AUTHOR:                                                      *
+ *       Sridhar Voruganti - sridhar.voruganti@modusbox.com               *
+ **************************************************************************/
+
+'use strict';
+
```
Comment (eoln) [inline src/lib/model/OutboundThirdpartyTransactionModel.js:12]: I have asked to refactor the publication to channel code, so maybe putting it in this channel is a good idea.

File: src/lib/model/InboundThirdpartyTransactionModel.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            // project our internal authorizations reponse into mojaloop response form
+            const mojaloopResponse = shared.internalAuthorizationsResponseToMojaloop(response);
+            // make a callback to the source fsp with the party info
+            return this._mojaloopRequests.putAuthorizations(authorizationsReq.transactionRequestId, mojaloopResponse, sourceFspId);
+        }
+        catch (err) {
+            this._logger.push({ err }).log('Error in postAuthorizations');
+            const mojaloopError = await this._handleError(err);
+            this._logger.push({ mojaloopError }).log(`Sending error response to ${sourceFspId}`);
+            return await this._mojaloopRequests.putAuthorizationsError(authorizationsReq.transactionRequestId, mojaloopError, sourceFspId);
```
Comment (eoln) [inline src/lib/model/InboundThirdpartyTransactionModel.js:86]: when you are returning from async function there is no need to use the `await` keyword

File: src/lib/model/InboundThirdpartyTransactionModel.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    async postAuthorizations(authorizationsReq, sourceFspId) {
+        try {
+            //Converts a mojaloop authorizationsReq data to internal form
+            const internalForm = shared.mojaloopAuthorizationsReqToInternal(authorizationsReq);
+
+            // make a call to the backend to ask for a authorizations response
+            const response = await this._backendRequests.getSignedChallenge(internalForm);
+            if (!response) {
+                // make an error callback to the source fsp
+                return 'No response from backend';
```
Comment (eoln) [inline src/lib/model/InboundThirdpartyTransactionModel.js:75]: shouldn't be a `throw` instead `return` ?

File: src/test/unit/inboundApi/handlers.test.js [hunk 11]
```diff
@@ -63,15 +67,15 @@ describe('Inbound API handlers:', () => {
             expect(quoteRequestSpy.mock.calls[0][1]).toBe(mockContext.request.headers['fspiop-source']);
```
Comment (eoln) [inline src/test/unit/inboundApi/handlers.test.js:67]: is possible to use `toHaveBeenCalledWith` instead mock.calls[...][...] ?

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+const Cache = require('@internal/cache');
+
+describe('inboundThirdpartyTransactionModel', () => {
+    let config;
+    let mockAuthReqArgs;
+    let logger;
+
+    beforeAll(async () => {
+        const logTransports = await Promise.all([Transports.consoleDir()]);
+        logger = new Logger({
```
Comment (eoln) [inline src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js:31]: can you use mockLogger here?

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            cache = new Cache({
+                host: 'dummycachehost',
+                port: 1234,
+                logger,
+            });
+            await cache.connect();
+
+            model = new ThirdpartyTrxnModelIn({
+                ...config,
+                cache,
```
Comment (eoln) [inline src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js:58]: there is no need to used full cache with mocked redis-> instead mock only used, by your code, cache's methods

File: src/InboundServer/handlers.js [hunk 14]
```diff
@@ -58,6 +60,50 @@ const getAuthorizationsById = async (ctx) => {
     ctx.response.body = '';
 };
 
+/**
+ * Handles POST /authorizations request
+ */
+const postAuthorizations = async (ctx) => {
+    // kick off an asyncronous operation to handle the request
+    (async () => {
```
[the remark this answers is not in the document]
Comment (sridharvoruganti) [reply src/InboundServer/handlers.js:69]: followed existing pattern used for other endpoints

File: src/InboundServer/handlers.js [hunk 15]
```diff
@@ -58,6 +60,50 @@ const getAuthorizationsById = async (ctx) => {
...
+
+            // use the model to handle the request
+            const response = await thirdpartyTrxnModelIn.postAuthorizations(ctx.request.body, sourceFspId);
+
+            // log the result
+            ctx.state.logger.push({ response }).log('Inbound Third party transaction model handled POST /authorizations request');
+        }
+        catch (err) {
+            // nothing we can do if an error gets thrown back to us here apart from log it and continue
+            ctx.state.logger.push({ err }).log('Error handling POST /authorizations}');
```
[the remark this answers is not in the document]
Comment (sridharvoruganti) [reply src/InboundServer/handlers.js:98]: we are handling errors in model, and calling PUT /authorizations/{ID}/error from that method

File: src/test/unit/inboundApi/handlers.test.js [hunk 16]
```diff
@@ -47,10 +51,10 @@ describe('Inbound API handlers:', () => {
                 state: {
                     conf: {},
                     // example of elaborative logging with keepQuite = false
```
[the remark this answers is not in the document]
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js]: Ok removed that

File: src/test/unit/inboundApi/handlers.test.js [hunk 17]
```diff
@@ -98,12 +102,44 @@ describe('Inbound API handlers:', () => {
...
+                response: {},
+                state: {
+                    conf: {},
+                    logger: mockLogger({ app: 'inbound-handlers-unit-test' })
+                }
+            };
+        });
+
+        test('calls `ThirdpartyTrxnModelIn.postAuthorizations` with the expected arguments.', async () => {
+            const authorizationRequestSpy = jest.spyOn(ThirdpartyTrxnModelIn.prototype, 'postAuthorizations');
```
[the remark this answers is not in the document]
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js:126]: Ok tried, but without prototype not working for classes

File: src/test/unit/inboundApi/handlers.test.js [hunk 18]
```diff
@@ -98,12 +102,44 @@ describe('Inbound API handlers:', () => {
...
+
+            await expect(handlers['/authorizations'].post(mockAuthorizationContext)).resolves.toBe(undefined);
+
+            expect(authorizationRequestSpy).toHaveBeenCalledTimes(1);
+            expect(authorizationRequestSpy.mock.calls[0][0]).toBe(mockAuthorizationContext.request.body);
```
[the remark this answers is not in the document]
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js]: Ok fixed it

File: src/InboundServer/handlers.js (hunk 1 above)
Comment (sridharvoruganti) [reply src/InboundServer/handlers.js:15]: Ok

File: src/InboundServer/handlers.js (hunk 2 above)
Comment (sridharvoruganti) [reply src/InboundServer/handlers.js:654]: As discussed ignoring it for now

File: src/InboundServer/handlers.js (hunk 3 above)
Comment (sridharvoruganti) [reply src/InboundServer/handlers.js]: Ok

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 4 above)
Comment (sridharvoruganti) [reply src/lib/model/InboundThirdpartyTransactionModel.js]: Deleted, actually we will use same inbound and outbound models for 331, 332 stories, so I thought these properties might be required

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 5 above)
Comment (sridharvoruganti) [reply src/lib/model/InboundThirdpartyTransactionModel.js]: Deleted

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 6 above)
Comment (sridharvoruganti) [reply src/lib/model/InboundThirdpartyTransactionModel.js]: Deleted

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 7 above)
Comment (sridharvoruganti) [reply src/lib/model/InboundThirdpartyTransactionModel.js]: removed that comment. followed existing pattern

File: src/lib/model/OutboundThirdpartyTransactionModel.js (hunk 8 above)
Comment (sridharvoruganti) [reply src/lib/model/OutboundThirdpartyTransactionModel.js:12]: Ok

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 9 above)
Comment (sridharvoruganti) [reply src/lib/model/InboundThirdpartyTransactionModel.js:77]: Followed existing pattern.

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 10 above)
Comment (sridharvoruganti) [reply src/lib/model/InboundThirdpartyTransactionModel.js:66]: Followed existing pattern.

File: src/test/unit/inboundApi/handlers.test.js (hunk 11 above)
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js]: Fixed it

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js (hunk 12 above)
Comment (sridharvoruganti) [reply src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js]: Ok , Fixed it

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js (hunk 13 above)
Comment (sridharvoruganti) [reply src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js]: Fixed it

File: src/test/unit/inboundApi/handlers.test.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (eoln) [reply src/test/unit/inboundApi/handlers.test.js:126]: sorry, you are right!

File: src/lib/model/InboundThirdpartyTransactionModel.js (hunk 9 above)
Comment (eoln) [reply src/lib/model/InboundThirdpartyTransactionModel.js:77]: yeah, that is strange, but let keep the existing convention

File: src/lib/model/OutboundThirdpartyTransactionModel.js [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+        throw new Error('OutboundThirdpartyTransactionModel.notificationChannel: \'id\' parameter is required');
+    }
+
+    // channel name
+    return `3ptrxnreq_${id}`;
+}
+
+async function publishNotifications(cache, id, value) {
+    const channel = notificationChannel(id);
+    cache.publish(channel, value);
```
Comment (eoln) [inline src/lib/model/OutboundThirdpartyTransactionModel.js:25]:
```suggestion
    return cache.publish(channel, value);
```

File: src/test/unit/inboundApi/handlers.test.js [hunk 20]
```diff
@@ -17,24 +18,28 @@ const handlers = require('../../../InboundServer/handlers');
...
 
 
 describe('Inbound API handlers:', () => {
     let mockArgs;
     let mockTransactionRequest;
+    let mockAuthReqArgs;
 
     beforeEach(() => {
         mockArgs = JSON.parse(JSON.stringify(mockArguments));
         mockTransactionRequest = JSON.parse(JSON.stringify(mockTransactionRequestData));
```
Comment (eoln) [inline src/test/unit/inboundApi/handlers.test.js:35]: let define `function deepClone(obj) { return JSON.parse(JSON.stringify(obj)); }` and use it

File: src/test/unit/inboundApi/handlers.test.js [hunk 21]
```diff
@@ -59,19 +63,18 @@ describe('Inbound API handlers:', () => {
             await expect(handlers['/quotes'].post(mockContext)).resolves.toBe(undefined);
 
             expect(quoteRequestSpy).toHaveBeenCalledTimes(1);
-            expect(quoteRequestSpy.mock.calls[0][0]).toBe(mockContext.request.body);
-            expect(quoteRequestSpy.mock.calls[0][1]).toBe(mockContext.request.headers['fspiop-source']);
+            expect(quoteRequestSpy).toHaveBeenCalledWith(mockContext.request.body, mockContext.request.headers['fspiop-source']);
```
Comment (eoln) [inline src/test/unit/inboundApi/handlers.test.js:66]:
```suggestion
            expect(quoteRequestSpy).toHaveBeenCalledWith(
                mockContext.request.body,
                mockContext.request.headers['fspiop-source']
            );
```

Comment (eoln) [reply src/test/unit/inboundApi/handlers.test.js:66]: let keep line width short

File: src/test/unit/inboundApi/handlers.test.js [hunk 22]
```diff
@@ -93,17 +96,48 @@ describe('Inbound API handlers:', () => {
...
+                }
+            };
+        });
+
+        test('calls `ThirdpartyTrxnModelIn.postAuthorizations` with the expected arguments.', async () => {
+            const authorizationRequestSpy = jest.spyOn(ThirdpartyTrxnModelIn.prototype, 'postAuthorizations');
+
+            await expect(handlers['/authorizations'].post(mockAuthorizationContext)).resolves.toBe(undefined);
+            expect(authorizationRequestSpy).toHaveBeenCalledTimes(1);
+            expect(authorizationRequestSpy).toHaveBeenCalledWith(mockAuthorizationContext.request.body,
```
Comment (eoln) [inline src/test/unit/inboundApi/handlers.test.js:130]: please reformat to keep line width short

File: src/test/unit/inboundApi/handlers.test.js [hunk 23]
```diff
@@ -129,15 +163,16 @@ describe('Inbound API handlers:', () => {
             await expect(handlers['/authorizations/{ID}'].get(mockAuthorizationContext)).resolves.toBe(undefined);
 
             expect(authorizationsSpy).toHaveBeenCalledTimes(1);
-            expect(authorizationsSpy.mock.calls[0][1]).toBe(mockAuthorizationContext.request.headers['fspiop-source']);
+            expect(authorizationsSpy).toHaveBeenCalledWith(mockAuthorizationContext.state.path.params.ID,
+                mockAuthorizationContext.request.headers['fspiop-source']);
```
Comment (eoln) [inline src/test/unit/inboundApi/handlers.test.js:167]: please reformat

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+        });
+
+        afterEach(async () => {
+            MojaloopRequests.__putAuthorizations.mockClear();
+        });
+
+        test('calls `mojaloopRequests.putAuthorizations` with the expected arguments.', async () => {
+            await model.postAuthorizations(mockAuthReqArgs.authorizationRequest, mockAuthReqArgs.fspId);
+            expect(MojaloopRequests.__putAuthorizations).toHaveBeenCalledTimes(1);
+            expect(MojaloopRequests.__putAuthorizations).toHaveBeenCalledWith(mockAuthReqArgs.authorizationRequest.transactionRequestId,
```
Comment (eoln) [inline src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js:52]: please reformat

File: src/lib/model/OutboundThirdpartyTransactionModel.js (hunk 19 above)
Comment (sridharvoruganti) [reply src/lib/model/OutboundThirdpartyTransactionModel.js]: Ok

File: src/test/unit/inboundApi/handlers.test.js (hunk 20 above)
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js]: Ok

File: src/test/unit/inboundApi/handlers.test.js (hunk 21 above)
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js]: Ok formatted now

File: src/test/unit/inboundApi/handlers.test.js (hunk 22 above)
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js]: Ok formatted now

File: src/test/unit/inboundApi/handlers.test.js (hunk 23 above)
Comment (sridharvoruganti) [reply src/test/unit/inboundApi/handlers.test.js:172]: Ok formatted now

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js (hunk 24 above)
Comment (sridharvoruganti) [reply src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js]: Ok formatted now

File: src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+const { BackendRequests } = require('@internal/requests');
+const mockLogger = require('../../mockLogger');
+
+describe('inboundThirdpartyTransactionModel', () => {
+    let config;
+    let mockAuthReqArgs;
+    let logger;
+
+    beforeEach(async () => {
+        config = JSON.parse(JSON.stringify(defaultConfig));
```
Comment (eoln) [inline src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js:28]:
```suggestion
        config = deepClone(defaultConfig);
```

Comment (sridharvoruganti) [reply src/test/unit/lib/model/InboundThirdpartyTransactionModel.test.js]: Ok it moved to deepClone()

**Pull request merged**